File: googleSpeedRun/170425/ransomNote.js
/* Given two strings ransomNote and magazine, return true if ransomNote can be 
constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.

Example 1:

Input: ransomNote = "a", magazine = "b"
Output: false
Example 2:

Input: ransomNote = "aa", magazine = "ab"
Output: false
Example 3:

Input: ransomNote = "aa", magazine = "aab"
Output: true */
 
const canConstruct = (ransomNote, magazine) => {

    let magazineMap = new Map();
    for(let char of magazine) magazineMap.set(char, (magazineMap.get(char) || 0) + 1);
    for (let char of ransomNote) {
        if(!magazineMap.has(char) || magazineMap.get(char) === 0) return false;
        //usamos una letrita, ahora la quitamos para evitar bugs
        magazineMap.set(char, (magazineMap.get(char) - 1));
    }
    return true;
};

console.log(canConstruct("aa", "ab"));

File: python/roadmap/hashmaps & sets/ransomNote.py
""" Given two strings ransomNote and magazine, return true if ransomNote 
can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.

Example 1:
Input: ransomNote = "a", magazine = "b"
Output: false

Example 2:
Input: ransomNote = "aa", magazine = "ab"
Output: false

Example 3:
Input: ransomNote = "aa", magazine = "aab"
Output: true """

from typing import Counter

def canConstruct(ransomNote: str, magazine: str) -> bool:

    #an empty ransomNote can always be constructed
    if not ransomNote:
        return True

    hashMap = {}
    for ch in magazine:
        hashMap[ch] = hashMap.get(ch, 0) + 1
    
    for ch in ransomNote:
        if hashMap.get(ch, 0) <= 0:
            return False
        hashMap[ch] -= 1
    
    return True

def canConstruct2(ransomNote:str, magazine: str) -> bool:
    
    hashMap = Counter(magazine)

    for ch in ransomNote:
        if hashMap[ch] > 0:
            hashMap[ch] -= 1
        else:
            return False
    
    return True

#print(canConstruct("aa", "aab"))
print(canConstruct2("aa", "aab"))
